game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/EvaderRetreat.java

```diff
@@ -0,0 +1,143 @@
+package games.strategy.triplea.delegate.battle.steps.retreat;
+
+import games.strategy.engine.data.CompositeChange;
+import games.strategy.engine.data.GamePlayer;
+import games.strategy.engine.data.Territory;
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.data.changefactory.ChangeFactory;
+import games.strategy.engine.delegate.IDelegateBridge;
+import games.strategy.triplea.delegate.battle.BattleActions;
+import games.strategy.triplea.delegate.battle.BattleState;
+import games.strategy.triplea.delegate.battle.MustFightBattle;
+import games.strategy.triplea.formatter.MyFormatter;
+import java.util.ArrayList;
+import java.util.Collection;
+import lombok.Builder;
+import lombok.experimental.UtilityClass;
+import org.triplea.sound.SoundUtils;
+
+@UtilityClass
+public class EvaderRetreat {
+
+  @Builder(toBuilder = true)
+  static class Parameters {
+    BattleState battleState;
+    BattleActions battleActions;
+    BattleState.Side side;
+    IDelegateBridge bridge;
+    Collection<Territory> possibleRetreatSites;
+    Collection<Unit> units;
+    String step;
+  }
+
+  public static void retreatUnits(final Parameters parameters) {
+    final BattleState battleState = parameters.battleState;
+    final BattleActions battleActions = parameters.battleActions;
+    final BattleState.Side side = parameters.side;
+    final IDelegateBridge bridge = parameters.bridge;
+    final Collection<Territory> possibleRetreatSites = parameters.possibleRetreatSites;
+    final Collection<Unit> units = parameters.units;
+    final String step = parameters.step;
+    final GamePlayer retreatingPlayer =
+        side == BattleState.Side.DEFENSE ? battleState.getDefender() : battleState.getAttacker();
+    final String text = retreatingPlayer.getName() + " retreat subs?";
+
+    bridge.getDisplayChannelBroadcaster().gotoBattleStep(battleState.getBattleId(), step);
+    final boolean isAttemptingSubmerge =
+        possibleRetreatSites.size() == 1
+            && possibleRetreatSites.contains(battleState.getBattleSite());
+    final Territory retreatTo =
+        isAttemptingSubmerge
+            ? battleActions.querySubmergeTerritory(
+                battleState, bridge, retreatingPlayer, possibleRetreatSites, text)
+            : battleActions.queryRetreatTerritory(
+                battleState, bridge, retreatingPlayer, possibleRetreatSites, text);
+    if (retreatTo == null) {
+      return;
+    }
+    playSound(battleState, bridge, retreatingPlayer, units);
+    if (battleState.getBattleSite().equals(retreatTo)) {
+      submergeEvaders(battleState, battleActions, units, side, bridge);
+      broadcastRetreat(bridge, retreatingPlayer, step, " submerges subs");
+    } else {
+
+      final CompositeChange change = new CompositeChange();
```

Comment:
nit, mixture of abstractions code smell here. The other methods are high level, then this bit of functionality is very low level. Recommend to extract to a method so we can keep a consistent level of abstraction, eG: "updateHistoryWithUnitsMoved(..)"